seed/seedwriter/seed16.go

```diff
@@ -0,0 +1,184 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2014-2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package seedwriter
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/seed"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/naming"
+)
+
+type policy16 struct {
+	model *asserts.Model
+	opts  *Options
+
+	needsCore   []string
+	needsCore16 []string
+}
+
+func (pol *policy16) systemSnap() *asserts.ModelSnap {
+	if pol.model.Classic() {
+		// no predefined system snap, infer later
+		return nil
+	}
+	snapName := "core"
+	snapType := "core"
+	if pol.model.Base() != "" {
+		snapName = "snapd"
+		snapType = "snapd"
+	}
+	// TODO: set SnapID too
+	return &asserts.ModelSnap{
+		Name:           snapName,
+		SnapType:       snapType,
+		Modes:          []string{"run"},
+		DefaultChannel: "stable",
+		Presence:       "required",
+	}
+}
+
+func (pol *policy16) checkBase(info *snap.Info, availableSnaps *naming.SnapSet) error {
+	// Sanity check, note that we could support this case
+	// if we have a use-case but it requires changes in the
+	// devicestate/firstboot.go ordering code.
+	if info.GetType() == snap.TypeGadget && !pol.model.Classic() && info.Base != pol.model.Base() {
```

Comment:
Gadget may not have a base, no? Or is it only the case with classic model? Should we say `... && info.Base != "" && info.Base != pol.model.Base() {`?

Reply:
No, so far gadget always have a base, same base as the model (on Core), so info.Base == "" means "core" as for apps.  This code is actually the same as code we have already in image.go: